resharper/resharper-unity/src/Yaml/Psi/Caches/UnityEditorPropertyValues/UnitySceneDataLocalCache.cs

```diff
@@ -0,0 +1,289 @@
+                } else 
+                {
+                    ProcessSceneHierarchyFromComponentToRootInner(sourceFile, gameObject, consumer, null);
+                }
+            }
+            else
+            {
+                ProcessSceneHierarchyFromComponentToRootInner(sourceFile, start, consumer, null);
+            }
+        }
+        
+        
+        // Invariant : startGameObject is Transform Component if it is not stripped 
+        // This method traverse scene hierarchy via visiting transform components and push corresponding to transform GameObject into consumer
+        private void ProcessSceneHierarchyFromComponentToRootInner(IPsiSourceFile currentFile, IUnityHierarchyElement startUnityObject,
+            IUnityCachedSceneProcessorConsumer consumer, ModificationHierarchyElement modifications)
+        {
+            
+            var currentUnityObject = startUnityObject;
+            while (currentUnityObject != null)
+            {
+                // Unity object could be stripped, it means, that corresponding real object belongs to another yaml file
+                // Also, it has reference to prefab instance in current file, which stores all prefab modification
+                if (currentUnityObject.IsStripped)
+                {
+                    var correspondingId = currentUnityObject.CorrespondingSourceObject;
+                    var prefabInstanceId = currentUnityObject.PrefabInstance;
+                    
+                    
+                    
+                    if (correspondingId == null || prefabInstanceId == null)
+                        return;
+
+                    var prefabInstance = mySceneElements.GetValues(new SceneElementId(currentFile, prefabInstanceId)).FirstOrDefault() as ModificationHierarchyElement;
+                    var prefabSourceFilePaths = myGuidCache.GetAssetFilePathsFromGuid(correspondingId.guid);
+                    if (prefabSourceFilePaths.Count > 1 || prefabSourceFilePaths.Count == 0)
+                        return;
+
+                    var unityModule = currentFile.PsiModule as UnityExternalFilesPsiModule;
+                    if (unityModule == null)
+                        return;
+                    
+                    if (!unityModule.NotNull("externalFilesModuleFactory.PsiModule != null")
+                        .TryGetFileByPath(prefabSourceFilePaths.First(), out var prefabSourceFile))
+                        return;
+                        
+
+                    var prefabStartGameObject = mySceneElements.GetValues(new SceneElementId(prefabSourceFile, correspondingId.WithGuid(null))).FirstOrDefault();
+                    
+                    if (prefabStartGameObject == null)
+                        return; // TODO [vkrasnotsvetov] 19.3 Handle case, when prefab contains prefab which contains prefab
+                    
+                    if (!prefabStartGameObject.IsStripped)
```

Comment:
can this piece of code be simplified? I.t. ifs can be probably merged and apparently the whole code arrangement could be done more nicely